Embedded Software/TelemetryControlUnit/src/logger.cpp

```diff
@@ -0,0 +1,250 @@
+uint32_t lastLogTime;
+uint32_t minLogFrequency;
+uint32_t startUpTimeMillis;
+uint32_t startUpTimeRTC;
+
+
+/**
+ * @brief Wrapper function to pass to time sync function
+ * 
+ * @return time_t 
+ */
+static time_t getTeensy3Time()
+{
+  return Teensy3Clock.get();
+}
+
+
+/**
+ * @brief Sets the logging file name using the format log-NUM.txt, where NUM is either the current 
+ *        unix time if using the RTC or a valid numeric index starting from 0.
+ * 
+ */
+static void setFileName() {
+  uint32_t fileNum = 0;
+  if (startUpTimeRTC > 0) {
+      fileNum = startUpTimeRTC;
+  }
+
+  snprintf(fileName, 19, "log-%lu.txt\0", fileNum);
+
+  while (SD.exists(fileName)) {
+    fileNum++;
+    snprintf(fileName, 19, "log-%lu.txt\0", fileNum); 
+  }
+}
+
+
+/**
+ * @brief Resets the buffers, SD card, and clock
+ * 
+ */
+static void reset() {
+  initialized = false;
+  buf1Length = 0;
+  buf2Length = 0;
+  memset((void *)messageBuf1, 0, MAX_BUFFERED_MESSAGES * sizeof(message_format_t));
+  memset((void *)messageBuf2, 0, MAX_BUFFERED_MESSAGES * sizeof(message_format_t));
+  usingBuf1 = true;
+}
+
+
+/**
+ * @brief Get the date/time in the format YYYY-MM-DDT00:00:00.000Z
+ * 
+ * @param timestamp Char pointer to hold return value
+ */
+static void getTimestamp(char *timestamp) {
+  uint32_t millisSinceStart = millis() - startUpTimeMillis;
+  uint32_t currentTimeSec = startUpTimeRTC + (millisSinceStart / 1000);
+  uint32_t extraMillis = millisSinceStart % 1000;
+
+  snprintf(timestamp, 25, "%.4d-%.2d-%.2dT%.2d:%.2d:%.2d.%.3luZ\0", year(currentTimeSec), month(currentTimeSec), 
+          day(currentTimeSec), hour(currentTimeSec), minute(currentTimeSec), second(currentTimeSec), extraMillis);
+}
+
+
+/**
+ * @brief Initilizes SD logging
+ * 
+ */
+int LoggerInit(uint32_t logFrequency) {
+  // Start built-in SD card
+  if (!SD.begin(BUILTIN_SDCARD)) {
+    DPRINTLN(F("Error initializing SD card logging"));
+    return LOGGER_ERROR_SD_CARD;
+  }
+
+  // Init the RTC
+  setSyncProvider(getTeensy3Time);   // the function to get the time from the RTC
+  
+  // Init startup times
+  if (timeStatus() == timeSet) {
+    DPRINTLN(F("RTC has set the system time")); 
+    startUpTimeRTC = now();
+  } else {
+    DPRINTLN(F("Unable to sync with the RTC"));
+    startUpTimeRTC = 0;
+  }
+  startUpTimeMillis = millis();
+  
+  // set the file name 
+  setFileName();
+  DPRINT(F("Setup complete. File name is "));
+  DPRINTLN(fileName);
+
+  minLogFrequency = logFrequency;
+  lastLogTime = millis();
+  logFile = SD.open(fileName, FILE_WRITE);
+  
+  initialized = true;
+  return LOGGER_SUCCESS;
+}
+
+
+/**
+ * @brief Writes the messages currently buffered in messageBuf to the SD card
+ * 
+ * @return true true on a successful write 
+ * @return false when the write fails
+ */
+int LoggerWrite() {
+  if (!initialized) {
+    return LOGGER_ERROR_SD_CARD;
+  }
+
+  // find appropriate buffer to use
+  message_format_t *messageBuf = usingBuf1 ? messageBuf1 : messageBuf2; 
```

Comment:
Why do we need two messageBufs?

Reply:
Potential scenario with only one buffer: 
- It fills up which triggers a write to the SD card to occur
- While the SD Write is happening, the program is interrupted from an incoming CAN message
- Program writes this message to the one buffer currently being logged to the SD card, either overwriting current data or exceeding the buffer capacity
Two buffers fixes this as whenever one fills up, we set the other to be the active receive buffer.